EventFilter/GEMRawToDigi/plugins/GEMDigiToRawModule.cc

```diff
@@ -67,37 +69,63 @@ void GEMDigiToRawModule::produce(edm::StreamID iID, edm::Event& iEvent, edm::Eve
   std::vector<std::unique_ptr<AMC13Event>> amc13Events;
   amc13Events.reserve(FEDNumbering::MAXGEMFEDID - FEDNumbering::MINGEMFEDID + 1);
 
+  uint32_t LV1_id = iEvent.id().event();
+  uint16_t BX_id = iEvent.bunchCrossing();
+  uint32_t OrN = iEvent.orbitNumber();
+
+  // making map of bx GEMDigiCollection
+  // each bx will be saved as new AMC13Event, so GEMDigiCollection needs to be split into bx
+  std::map<int, GEMDigiCollection> gemBxMap;
+  for (auto const& etaPart : *gemDigis) {
+    GEMDetId gemId = etaPart.first;
+    const GEMDigiCollection::Range& digis = etaPart.second;
+    for (auto digi = digis.first; digi != digis.second; ++digi) {
+      int bx = digi->bx();
+      auto search = gemBxMap.find(bx);
+      if (search != gemBxMap.end()) {
+        search->second.insertDigi(gemId, *digi);
+      } else {
+        GEMDigiCollection newGDC;
+        newGDC.insertDigi(gemId, *digi);
+        gemBxMap.insert(std::pair<int, GEMDigiCollection>(bx, newGDC));
+      }
+    }
+  }
+
+  uint32_t amc13EvtLength = 0;
   for (unsigned int fedId = FEDNumbering::MINGEMFEDID; fedId <= FEDNumbering::MAXGEMFEDID; ++fedId) {
     std::unique_ptr<AMC13Event> amc13Event = std::make_unique<AMC13Event>();
 
-    for (uint8_t amcNum = 0; amcNum < GEMeMap::maxAMCs_; ++amcNum) {
-      std::unique_ptr<AMCdata> amcData = std::make_unique<AMCdata>();
+    for (auto const& gemBx : gemBxMap) {
+      int bx = gemBx.first;
+      GEMDigiCollection newGemDigis = gemBx.second;
```

Comment:
perhaps some more descriptive name rather then ```newGemDigis``` can be found, e.g. ```inBxGemDigis```

Reply:
done